torchrl/data/rlhf/utils.py

```diff
@@ -0,0 +1,321 @@
+              each time step
+            - "sample_log_prob": the log probability of each token during generation
+            - ("next", "input_ids"): the sequence of tokens after generation. Makes up
+              part of the inputs that will be used for generating the next token.
+            - ("next", "attention_mask"): updated attention_mask after token has been
+              generated. Passed to the generative model on the next time step
+            - ("next", "done"): Boolean array indicating whether we've reached a
+              terminal state (either because we generated EOS token or because we
+              reached the token limit)
+            - ("next", "reward"): The reward received at each time step
+            - ("next", "reward_raw"): The raw reward from the reward model, without the
+              KL term. This is mainly for debugging and logging, it is not used in
+              training
+            - ("next", "reward_kl"): The KL term from the reward. This is mainly for
+              debugging and logging, it is not used in training.
+        """
+        rollout_generated = []
+        for rindex, row in zip(batch.prompt_rindex, generated):
+            arange = torch.arange(row.shape[0], device=generated.device)
+            tokens = []
+            for i in range(self.max_new_tokens + 1):
+                tokens.append(
+                    torch.where(
+                        arange < rindex + i,
+                        row,
+                        self.EOS_TOKEN_ID,
+                    )
+                )
+            rollout_generated.append(torch.stack(tokens))
+        rollout_generated = torch.stack(rollout_generated)
+        rollout_attention_mask = (rollout_generated != self.EOS_TOKEN_ID).bool()
+
+        # done is True when we either first sample an EOS token or reach the maximum number
+        # of generated tokens
+        done_idx = torch.minimum(
+            (generated != self.EOS_TOKEN_ID).sum(dim=-1) - batch.prompt_rindex,
+            torch.tensor(self.max_new_tokens) - 1,
+        )
+        done = torch.zeros(
+            done_idx.numel(),
+            self.max_new_tokens,
+            dtype=torch.bool,
+            device=generated.device,
+        )
+        done = done.scatter(-1, done_idx.unsqueeze(-1), 1).unsqueeze(-1)
+
+        # the sequence of actions for each trajectory is just the generated token ids
+        action_idx = torch.arange(self.max_new_tokens, device=generated.device)
+        action_idx = action_idx + batch.prompt_rindex.unsqueeze(-1)
+        action = generated.gather(-1, action_idx)
+
+        # calculate the reward for the finished sequence
+        _, end_scores = self.reward_model(
+            input_ids=rollout_generated[:, -1],
+            attention_mask=rollout_attention_mask[:, -1],
+        )
+        _, end_scores_labels = self.reward_model(
+            input_ids=batch.input_ids,
+            attention_mask=batch.attention_mask,
+        )
```

Comment:
do we need 2 calls? Can't this be grouped?

Reply:
I tried grouping but it's non-trivial because they're different shapes. We could pad `batch` perhaps?